### PR TITLE
Add fine-grained port mirroring and associated tests 

### PR DESCRIPTION
**Pull request by gwacter**
These are the changes associated with the work done at the Hackfest in Sydney (2016). Have tested against existing tests and added two additional tests for new ACL format and ACL mirror.

### CODE REVIEW

**Comment by anarkiwi** [comment]
Git reports merge conflicts.

Please resolve these conflicts and we can review again.

File: src/ryu_faucet/org/onfsdn/faucet/valve.py [hunk 1]
```diff
@@ -618,6 +627,8 @@ def port_delete(self, dp_id, port_num):
 
         # delete all rules matching this port in all tables.
         for table in self.all_valve_tables():
+            #if table == self.dp.acl_table:
+	    #  continue
```
Comment (KitL) [inline src/ryu_faucet/org/onfsdn/faucet/valve.py]: This shouldn't be commented right?

Comment (gwacter) [inline src/ryu_faucet/org/onfsdn/faucet/valve.py]: You are right -- it shouldn't be commented. We don't want to drop ACL rules unless the user requests the change.

**Comment by gwacter** [comment]
Conflicts removed, passing all tests except for MAX HOSTS.

**Comment by anarkiwi** [comment]
Great, thanks, looks better.

There are other changes to the base test class. Why are they necessary?

There is a magic constant: self.mirror_from_port[4294967040 - acl_num] =
port_no. Where does this number come from and why is it necessary?

This change attempts to maintain ACL rules between calls. I'm not opposed
to this in principle but let's not try to do this in this particular change
- we should focus solely on the new functionality added. Please remove that
code from this change.

**Comment by anarkiwi** [comment]
Thanks also for reporting the flakeyness in the MaxHosts test (it
occasionally fails in my setup also - OVS 2.4.1). I just committed a change
to the test to make it retry, if you'd like to merge that.

**Comment by anarkiwi** [comment]
Ping - is it possible to get these changes actioned please? With the requested changes the diff should be very small and small changes will be easy to verify and merge.

**Comment by Callum027** [comment]
Sorry for the late response, was not subscribed to this thread at first.

All tests seem to be passing on my end - @trungdtbk, could you please verify that everything is fine?

**Comment by trungdtbk** [comment]
Hi Callum,
I run the tests. Maxhosts and Mirrortest fail if I run all the tests together. But they are successful if run individually.

**Comment by anarkiwi** [comment]
Hi all;

Let's reset.

Trung - please can you see the comments earlier in this thread, about the
changes to the tests. We want to trim your change down to the minimal one
(I see for example, that you hard code OF1.3 in the test base class - why
is this is necessary)?

If there are changes needed to the base class, please add notes explaining
why to the pull request.

Thanks,

**Comment by trungdtbk** [comment]
Hi Josh,
I used mininet 2.2.0 and it fails to create OF1.3 switches without specifying 'OpenFlow13'. 

Regarding "4294967040" (=MAX of OF ports), I'm not so sure why we use that. But may be for avoiding collision because we use the same dict for both port mirroring and acl mirroring.

I've removed the codes at 707-708 regarding ACL rules.
Cheers,

**Comment by anarkiwi** [comment]
OK. We need to figure out why you need OF1.3 to be specified. It isn't
required in any other test scenario I know of. Which version of OVS are you
testing with?

We can't accept a change if we're not sure what the code does. :) What do
you intend to happen?

**Comment by trungdtbk** [comment]
I use ovs 2.0.2

Suppose if we remove the magic number, then the following config will create anomaly.
Port 1 has its traffic mirrored to port 2 by acl rule
Port 3 also has its traffic mirrored to port 2 (unwanted result)
interfaces:
     1:
            acl_in: 3
      2:
      3:
acls:
     3:
           -rule:
                  actions:
                         allow: 1
                         mirror: 2

**Comment by anarkiwi** [comment]
Per the comments in the test file, OVS 2.0.2 is unsupported.

Please use the recommend version and repeat the test.

**Comment by gwacter** [comment]
@trungdtbk  and I had a chat. We think that if we use the right version of OVS that we can remove the Openflow13 flag and return the print statements in faucet test back to the previous version (they were changed to work with the older version of OVS).

Regarding the magic number. It is a bit of hack to allow reuse of the mirror_from_port data structure belonging to dp so it can record both the case where all traffic is mirrored and the selective case. It works, might need refactoring in the longer term.

**Comment by anarkiwi** [comment]
Great (re OVS).

I'm sorry, we can't have the magic number hack. I'm happy to hear another
proposal (or discuss in person on Wednesday). Now that we are at V1.0 and
more and more people are relying on this code, we can't commit experimental
code to the codebase.

**Comment by gwacter** [comment]
I propose that we add an extra data structure to dp.py to do the housekeeping for fine grained mirroring. I think the only other change would be to modify the function **add_ports_and_vlans** to add any ports that are the target of those acls to the list of ports to be configured. 

@trungdtbk is going to work on that tomorrow.

**Comment by anarkiwi** [comment]
That sounds a lot cleaner. Great!

**Comment by gwacter** [comment]
@trungdtbk has simplified the code (removed extra data structure and optimisations re multiple drop flows).

**Comment by gwacter** [comment]
@trungdtbk just confirmed that it passes all the tests.

**Comment by anarkiwi** [comment]
I have merged it. This is a useful feature! I look forward to seeing it
used.

**Pull request merged**